controllers/src/main/java/io/syndesis/controllers/integration/IntegrationController.java

```diff
@@ -134,101 +162,132 @@ private void checkIntegrationStatus(Integration integration) {
-                            .statusMessage(Optional.ofNullable(update.getStatusMessage()))
-                            .stepsDone(Optional.ofNullable(update.getStepsPerformed()))
-                            .lastUpdated(new Date())
-                            .build());
                 }
 
             } catch (@SuppressWarnings("PMD.AvoidCatchingGenericException") Exception e) {
-                LOG.error("Error while processing integration status for integration {}", integrationId, e);
+                LOG.error("Error while processing integration status for integration {}", integration.getId().orElse("[none]"), e);
                 // Something went wrong.. lets note it.
-                Integration current = dataManager.fetch(Integration.class, integrationId);
-                dataManager.update(new Integration.Builder()
-                    .createFrom(current)
-                    .statusMessage("Error: "+e)
-                    .lastUpdated(new Date())
-                    .build());
-
+                updateRevisionStatus(integration, revision, new StatusUpdate(revision.getVersion().orElse(0), IntegrationState.Error, e.getMessage()));
             } finally {
                 // Add a next check for the next interval
-                reschedule(integrationId);
+                reschedule(integration.getId().get());
             }
-
         });
     }
 
+
+    /**
+     * Updates the {@link Integration} with the {@link StatusUpdate} returned by the {@link StatusChangeHandler}.
+     * @param integration      The integration.
+     * @param revision         The revision the update refers to.
+     * @param update           The status update.
+     */
+    private void updateRevisionStatus(Integration integration, IntegrationRevision revision, StatusUpdate update) {
+        Integration current = dataManager.fetch(Integration.class, integration.getId().get());
+
+        //In most cases status updates are about deployed/older revisions, with just a few exceptions.
+        // In all other cases we can safely discard the draft.
+        boolean retainDraft = integration.getDraftRevision().isPresent()
+            && (update.getState() == IntegrationState.Draft
+                || update.getState() == IntegrationState.Pending
+                || update.getState() == IntegrationState.Error);
+
+        IntegrationRevision updatedDraft = retainDraft
+            ? integration.getDraftRevision().get().withCurrentState(update.getState(), update.getStatusMessage())
+            : null;
+
+        //We need to iterate all revisions and update the status where there revision number matches.
+        List<IntegrationRevision> updatedRevisions = integration.getRevisions().stream()
+            .map(r -> r.getVersion().orElse(0).equals(revision.getVersion().equals(0))
+                ? r.withCurrentState(update.getState(), update.getStatusMessage())
+                : r)
+            .collect(Collectors.toList());
+
```

Comment:
What I'm missing here is, when a draft integration gets deployed, it should be added to the list of deployedRevisions. Could be that I missed that, but I thought this would be the correct place here ?